src/index.js

```diff
@@ -750,65 +748,94 @@ export default function (babel) {
 
     path.traverse({
 
-      ForFromArrayStatement(path) {
-        let init, test, update;
+      ForInArrayStatement(path) {
+        let refId = path.node.array;
 
-        let id = path.node.id;
-        init = t.variableDeclaration("let", [
-          t.variableDeclarator(id, t.numericLiteral(0))
-        ]);
-
-        test = t.binaryExpression("<", id,
-          t.memberExpression(path.node.array, t.identifier("length")));
-        update = t.updateExpression("++", id);
+        // If array is a complex expression, generate: const _array = expr
+        if (path.node.array.type !== "Identifier") {
+          refId = path.scope.generateUidIdentifier("array");
+          path.insertBefore(
+            t.variableDeclaration("const", [ t.variableDeclarator(refId, path.node.array )])
+          );
+        }
 
+        // let _i = 0, _len = array.length
+        let idx = path.node.idx ? path.node.idx : path.scope.generateUidIdentifier("i");
+        let len = path.scope.generateUidIdentifier("len");
+        let declarations = [
+          t.variableDeclarator(idx, t.numericLiteral(0)),
+          t.variableDeclarator(
+            len,
+            t.memberExpression(refId, t.identifier("length"))
+          )
+        ];
+        let init = t.variableDeclaration("let", declarations);
+        // _i < _len
+        let test = t.binaryExpression("<", idx, len);
+        // _i++
+        let update = t.updateExpression("++", idx);
+
+        // Element initializer: const elem = _array[_i]
         if (path.node.elem) {
-          // generate `const x = arr[i];`
-          let idPath = path.get("id");
-          let elemDecl = t.variableDeclaration("const", [
-            t.variableDeclarator(path.node.elem, t.memberExpression(
-              path.node.array, id, true
-            ))
-          ]);
-
-          if (t.isBlockStatement(path.node.body)) {
-            path.get("body").unshiftContainer("body", elemDecl);
-          } else {
-            path.get("body").replaceWith(t.blockStatement([
-              elemDecl,
-              path.node.body,
-            ]));
-          }
-          // may not be necessary?
-          path.get("body").scope.registerBinding("const", idPath);
+          path.scope.push({
```

Comment:
I read the `babel-handbook` to learn this stuff, and path.scope.push is what they recommend. I definitely think we should use it whenever it actually makes the code we want. I couldn't use it in the object case because it does hoist things all the way to the top of scope, but that's fine in the array case.